File: CS/GridControlAutoSize/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraEditors;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraGrid.Columns;
using System.Reflection;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Scrolling;
using DevExpress.XtraGrid.Views.Base;

namespace GridControlAutoSize {
    public partial class MainForm : XtraForm {
        private const int GridMinWidth = 400;
        private const int GridMinHeight = 200;

        public MainForm() {
            InitializeComponent();
        }

        private void MainForm_Load(object sender, EventArgs e) {
            // TODO: This line of code loads data into the 'nwindDataSet.Orders' table. You can move, or remove it, as needed.
            this.ordersTableAdapter.Fill(this.nwindDataSet.Orders);
            gridView1.ActiveFilterString = "CustomerID = 'ALFKI'";
        }

        private void UpdateGridSize() {
             Size maxSize = new Size(ClientSize.Width, ClientSize.Height);
             gridControl1.Size = gridControl1.CalcBestSize(maxSize, false);
             gridControl1.MainView?.LayoutChanged();
             gridControl1.Size = gridControl1.CalcBestSize(maxSize, true);
             gridControl1.MainView?.LayoutChanged();
        }

        private void OnGridViewRowCountChanged(object sender, EventArgs e) {
            UpdateGridSize();
        }

        private void OnGridViewColumnWidthChanged(object sender, ColumnEventArgs e) {
            UpdateGridSize();
        }

        private void OnGridViewDragObjectDrop(object sender, DragObjectDropEventArgs e) {
            UpdateGridSize();
        }

        private void MainForm_Shown(object sender, EventArgs e) {
            gridView1.RowCountChanged += new EventHandler(OnGridViewRowCountChanged);
            gridView1.ColumnWidthChanged += new ColumnEventHandler(OnGridViewColumnWidthChanged);
            gridView1.DragObjectDrop += new DragObjectDropEventHandler(OnGridViewDragObjectDrop);
            UpdateGridSize();
        }
    }
}

File: VB/GridControlAutoSize/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraEditors
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid.Columns
Imports System.Reflection
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Scrolling
Imports DevExpress.XtraGrid.Views.Base

Namespace GridControlAutoSize
	Partial Public Class MainForm
		Inherits XtraForm

		Private Const GridMinWidth As Integer = 400
		Private Const GridMinHeight As Integer = 200

		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub MainForm_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			' TODO: This line of code loads data into the 'nwindDataSet.Orders' table. You can move, or remove it, as needed.
			Me.ordersTableAdapter.Fill(Me.nwindDataSet.Orders)
			gridView1.ActiveFilterString = "CustomerID = 'ALFKI'"
		End Sub

		Private Sub UpdateGridSize()
			Dim maxSize As New Size(ClientSize.Width, ClientSize.Height)
                        gridControl1.Size = gridControl1.CalcBestSize(maxSize, False)
                        gridControl1.MainView?.LayoutChanged()
                        gridControl1.Size = gridControl1.CalcBestSize(maxSize, True)
                        gridControl1.MainView?.LayoutChanged()
		End Sub

		Private Sub OnGridViewRowCountChanged(ByVal sender As Object, ByVal e As EventArgs)
			UpdateGridSize()
		End Sub

		Private Sub OnGridViewColumnWidthChanged(ByVal sender As Object, ByVal e As ColumnEventArgs)
			UpdateGridSize()
		End Sub

		Private Sub OnGridViewDragObjectDrop(ByVal sender As Object, ByVal e As DragObjectDropEventArgs)
			UpdateGridSize()
		End Sub

		Private Sub MainForm_Shown(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Shown
			AddHandler gridView1.RowCountChanged, AddressOf OnGridViewRowCountChanged
			AddHandler gridView1.ColumnWidthChanged, AddressOf OnGridViewColumnWidthChanged
			AddHandler gridView1.DragObjectDrop, AddressOf OnGridViewDragObjectDrop
			UpdateGridSize()
		End Sub
	End Class
End Namespace
